### PR TITLE
 West Virginia low-income earned income exclusion (subtractions)

### PR DESCRIPTION
**Pull request by yedeer12**
add wv_low_income_earned_income_exclusion
reference = (
        "https://code.wvlegislature.gov/11-21-10/"
        "https://tax.wv.gov/Documents/TaxForms/2021/it140.pdf#page=20"
        "https://tax.wv.gov/Documents/PIT/2022/PersonalIncomeTaxFormsAndInstructions.2022.pdf#page=27"
    )
image

### CODE REVIEW

File: policyengine_us/parameters/gov/states/wv/tax/income/subtractions/low_income_earned_income_exclusion/income_threshold.yaml [hunk 1]
```diff
@@ -0,0 +1,26 @@
+description: West Virginia qualifies filers for the  low-income earned income exclusion with income below this threshold.
+metadata:
+  period: year
+  unit: USD
+  label: West Virginia low-income earned income exclusion federal adjusted gross income threshold
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/wv/tax/income/subtractions/low_income_earned_income_exclusion/income_threshold.yaml:5]:
```suggestion
  label: West Virginia low-income earned income exclusion adjusted gross income threshold
```

File: policyengine_us/parameters/gov/states/wv/tax/income/subtractions/low_income_earned_income_exclusion/income_threshold.yaml [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+  - title: 2021 West Virginia Personal Income Tax Forms And Instructions-2021 FAMILY TAX CREDIT TABLES
+    href: https://tax.wv.gov/Documents/TaxForms/2021/it140.pdf#page=20
+  - title: 2022 West Virginia Personal Income Tax Forms And Instructions-2022 FAMILY TAX CREDIT TABLES
+    href: https://tax.wv.gov/Documents/PIT/2022/PersonalIncomeTaxFormsAndInstructions.2022.pdf#page=27
+
+  breakdown:
+    - filing_status
+
+SINGLE:
+  2022-01-01: 10_000
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/wv/tax/income/subtractions/low_income_earned_income_exclusion/income_threshold.yaml:18]: was this introduced in 2022? 
If no - need 2021 value 

We also need to include in the subtractions.ymal file

File: policyengine_us/variables/gov/states/wv/tax/income/subtractions/low_income_earned_income_exclusion/wv_low_income_earned_income_exclusion.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    reference = (
+        "https://code.wvlegislature.gov/11-21-10/"
+        "https://tax.wv.gov/Documents/TaxForms/2021/it140.pdf#page=20"
+        "https://tax.wv.gov/Documents/PIT/2022/PersonalIncomeTaxFormsAndInstructions.2022.pdf#page=27"
+    )
+
+    def formula(tax_unit, period, parameters):
+        federal_agi = tax_unit("adjusted_gross_income", period)
+        person = tax_unit.members
+        wv_earned_income = person("earned_income", period)
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/wv/tax/income/subtractions/low_income_earned_income_exclusion/wv_low_income_earned_income_exclusion.py:20]: is it just head and spouse earned income or all?

File: policyengine_us/variables/gov/states/wv/tax/income/subtractions/low_income_earned_income_exclusion/wv_low_income_earned_income_exclusion_eligible.py [hunk 4]
```diff
@@ -0,0 +1,25 @@
+from policyengine_us.model_api import *
+
+
+class wv_low_income_earned_income_exclusion_eligible(Variable):
+    value_type = bool
+    entity = TaxUnit
+    label = "Eligible for the West Virginia low-income earned income exclusion"
+    defined_for = StateCode.WV
+    unit = USD
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/wv/tax/income/subtractions/low_income_earned_income_exclusion/wv_low_income_earned_income_exclusion_eligible.py:9]:
```suggestion
```

File: policyengine_us/parameters/gov/states/wv/tax/income/subtractions/low_income_earned_income_exclusion/income_threshold.yaml [hunk 5]
```diff
@@ -0,0 +1,26 @@
+description: West Virginia qualifies filers for the  low-income earned income exclusion with income below this threshold.
+metadata:
+  period: year
+  unit: USD
+  label: West Virginia low-income earned income exclusion federal adjusted gross income threshold
+  reference:
+  - title: Code of West Virginia §11-21-10 (b)
+    href: https://code.wvlegislature.gov/11-21-10/
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/wv/tax/income/subtractions/low_income_earned_income_exclusion/income_threshold.yaml:8]: i dont think I can open the link

File: policyengine_us/variables/gov/states/wv/tax/income/subtractions/low_income_earned_income_exclusion/wv_low_income_earned_income_exclusion.py (hunk 3 above)
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/wv/tax/income/subtractions/low_income_earned_income_exclusion/wv_low_income_earned_income_exclusion.py:20]: we also need a tax_unit.sum to sum up the vectorized values of the earned income

Comment (PavelMakarchuk) [reply policyengine_us/variables/gov/states/wv/tax/income/subtractions/low_income_earned_income_exclusion/wv_low_income_earned_income_exclusion.py:20]: then we just need an:

add(tax_unit, period, ["earned_income"])

File: changelog_entry.yaml [hunk 6]
```diff
@@ -0,0 +1,5 @@
+- bump: minor
+  changes:
+    added:
+    - West Virginia low-income earned income exclusion.
+    - West Virginia low-income earned income exclusion eligible.
```
Comment (PavelMakarchuk) [inline changelog_entry.yaml:5]:
```suggestion
```

File: CHANGELOG.md
Comment (PavelMakarchuk) [inline CHANGELOG.md:1]: need a git pull upstream master to fix the issues here

File: policyengine_us/variables/gov/states/wv/tax/income/subtractions/low_income_earned_income_exclusion/wv_low_income_earned_income_exclusion.py
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/wv/tax/income/subtractions/low_income_earned_income_exclusion/wv_low_income_earned_income_exclusion.py:1]: rename the folder to `earned_income`

File: policyengine_us/parameters/gov/states/wv/tax/income/subtractions/low_income_earned_income_exclusion/income_threshold.yaml
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/wv/tax/income/subtractions/low_income_earned_income_exclusion/income_threshold.yaml:1]: rename the folder to low_income_earned_income

File: policyengine_us/parameters/gov/states/wv/tax/income/subtractions/subtractions.yaml
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/wv/tax/income/subtractions/subtractions.yaml:1]: I dont think we should delete this here unless we have 2 files that are the same

Comment (PavelMakarchuk) [reply policyengine_us/parameters/gov/states/wv/tax/income/subtractions/subtractions.yaml:1]: just need to add `wv_low_income_earned_income_exclusion` to the list

File: policyengine_us/variables/gov/states/wv/tax/income/subtractions/low_income_earned_income_exclusion/wv_low_income_earned_income_exclusion.py [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+        federal_agi = tax_unit("adjusted_gross_income", period)
+        wv_earned_income = add(tax_unit, period, ["earned_income"])
+        filing_status = tax_unit("filing_status", period)
+
+        p = parameters(
+            period
+        ).gov.states.wv.tax.income.subtractions.low_income_earned_income_exclusion
+
+        return min_(
+            federal_agi, wv_earned_income, p.income_threshold[filing_status]
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/wv/tax/income/subtractions/low_income_earned_income_exclusion/wv_low_income_earned_income_exclusion.py:27]: I think the min_ function only take two arguments unless we do [this](https://stackoverflow.com/questions/39277638/element-wise-minimum-of-multiple-vectors-in-numpy/39279912#39279912)

or we can chain them (two min functions)

**Review by PavelMakarchuk — APPROVED** [review]

File: policyengine_us/variables/gov/states/wv/tax/income/subtractions/low_income_earned_income/wv_low_income_earned_income_exclusion.py [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+
+    def formula(tax_unit, period, parameters):
+        federal_agi = tax_unit("adjusted_gross_income", period)
+        wv_earned_income = add(tax_unit, period, ["earned_income"])
+        filing_status = tax_unit("filing_status", period)
+
+        p = parameters(
+            period
+        ).gov.states.wv.tax.income.subtractions.low_income_earned_income
+        income_min = min_(federal_agi, wv_earned_income)
```
Comment (MaxGhenis) [inline policyengine_us/variables/gov/states/wv/tax/income/subtractions/low_income_earned_income/wv_low_income_earned_income_exclusion.py:25]:
```suggestion
        # Lesser of federal AGI, earned income, and exclusion cap.
        income_min = min_(federal_agi, wv_earned_income)
```

File: policyengine_us/variables/gov/states/wv/tax/income/subtractions/low_income_earned_income/wv_low_income_earned_income_exclusion.py [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+    definition_period = YEAR
+    reference = (
+        "https://code.wvlegislature.gov/11-21-10/"
+        "https://tax.wv.gov/Documents/TaxForms/2021/it140.pdf#page=20"
+        "https://tax.wv.gov/Documents/PIT/2022/PersonalIncomeTaxFormsAndInstructions.2022.pdf#page=27"
+    )
+
+    def formula(tax_unit, period, parameters):
+        federal_agi = tax_unit("adjusted_gross_income", period)
+        wv_earned_income = add(tax_unit, period, ["earned_income"])
```
Comment (MaxGhenis) [inline policyengine_us/variables/gov/states/wv/tax/income/subtractions/low_income_earned_income/wv_low_income_earned_income_exclusion.py:19]: not wv specific
```suggestion
        earned_income = tax_unit("tax_unit_earned_income", period)
```

**Review by MaxGhenis — CHANGES_REQUESTED** [review]

File: policyengine_us/parameters/gov/states/wv/tax/income/subtractions/low_income_earned_income/amount.yaml [hunk 10]
```diff
@@ -0,0 +1,24 @@
+description: West Virginia limits the low-income earned income exclusion below this amount.
+  period: year
+  unit: USD
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/wv/tax/income/subtractions/low_income_earned_income/amount.yaml:3]:
```suggestion
  unit: currency-USD
```

File: policyengine_us/parameters/gov/states/wv/tax/income/subtractions/low_income_earned_income/income_limit.yaml [hunk 11]
```diff
@@ -0,0 +1,24 @@
+description: West Virginia limits the low-income earned income exclusion to filers with federal adjusted gross income below this amount.  
+  period: year
+  unit: USD
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/wv/tax/income/subtractions/low_income_earned_income/income_limit.yaml:3]:
```suggestion
  unit: currency-USD
```

File: policyengine_us/parameters/gov/states/wv/tax/income/subtractions/low_income_earned_income/amount.yaml [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+  label: West Virginia low-income earned income exclusion low-income earned income exclusion limit   
+  reference:
+  - title: Code of West Virginia §11-21-10 (b)
+    href: https://code.wvlegislature.gov/11-21-10/
+  - title: 2021 West Virginia Personal Income Tax Forms And Instructions-2021 FAMILY TAX CREDIT TABLES
+    href: https://tax.wv.gov/Documents/TaxForms/2021/it140.pdf#page=20
+  - title: 2022 West Virginia Personal Income Tax Forms And Instructions-2022 FAMILY TAX CREDIT TABLES
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/wv/tax/income/subtractions/low_income_earned_income/amount.yaml:10]: indent the title and href

File: policyengine_us/parameters/gov/states/wv/tax/income/subtractions/low_income_earned_income/income_limit.yaml [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+  label: West Virginia low-income earned income exclusion AGI limit   
+  reference:
+  - title: Code of West Virginia §11-21-10 (b)
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/wv/tax/income/subtractions/low_income_earned_income/income_limit.yaml:6]: indent the title and href

File: policyengine_us/parameters/gov/states/wv/tax/income/subtractions/low_income_earned_income/income_limit.yaml [hunk 14]
```diff
@@ -0,0 +1,24 @@
+description: West Virginia limits the low-income earned income exclusion to filers with federal adjusted gross income below this amount.  
+  period: year
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/wv/tax/income/subtractions/low_income_earned_income/income_limit.yaml:2]:
```suggestion
metadata:
  period: year
```

File: policyengine_us/parameters/gov/states/wv/tax/income/subtractions/low_income_earned_income/amount.yaml [hunk 15]
```diff
@@ -0,0 +1,24 @@
+description: West Virginia limits the low-income earned income exclusion below this amount.
+  period: year
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/wv/tax/income/subtractions/low_income_earned_income/amount.yaml:2]:
```suggestion
metadata:
  period: year
```

File: policyengine_us/parameters/gov/states/wv/tax/income/subtractions/low_income_earned_income/amount.yaml [hunk 16]
```diff
@@ -0,0 +1,24 @@
+description: West Virginia limits the low-income earned income exclusion below this amount.
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/wv/tax/income/subtractions/low_income_earned_income/amount.yaml:1]:
```suggestion
description: West Virginia caps the low-income earned income exclusion to this amount of earned income.
```

File: policyengine_us/variables/gov/states/wv/tax/income/subtractions/low_income_earned_income/wv_low_income_earned_income_exclusion_eligible.py [hunk 17]
```diff
@@ -0,0 +1,24 @@
...
+
+    def formula(tax_unit, period, parameters):
+        federal_agi = tax_unit("adjusted_gross_income", period)
+        filing_status = tax_unit("filing_status", period)
+
+        p = parameters(
+            period
+        ).gov.states.wv.tax.income.subtractions.low_income_earned_income
+
+        return federal_agi <= p.income_threshold[filing_status]
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/wv/tax/income/subtractions/low_income_earned_income/wv_low_income_earned_income_exclusion_eligible.py:24]:
```suggestion
        return federal_agi <= p.income_limit[filing_status]
```

File: policyengine_us/parameters/gov/states/wv/tax/income/subtractions/low_income_earned_income/income_limit.yaml [hunk 18]
```diff
@@ -0,0 +1,25 @@
+description: West Virginia limits the low-income earned income exclusion to filers with federal adjusted gross income below this amount.  
+metadata:
+  period: year
+  unit: currency-USD  
+  label: West Virginia low-income earned income exclusion AGI limit
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/wv/tax/income/subtractions/low_income_earned_income/income_limit.yaml:5]:
```suggestion
  label: West Virginia low-income earned income exclusion income limit 
```

File: policyengine_us/parameters/gov/states/wv/tax/income/subtractions/low_income_earned_income/income_limit.yaml [hunk 19]
```diff
@@ -0,0 +1,25 @@
+description: West Virginia limits the low-income earned income exclusion to filers with federal adjusted gross income below this amount.  
+metadata:
+  period: year
+  unit: currency-USD  
+  label: West Virginia low-income earned income exclusion AGI limit   
+  reference:
+    - title: Code of West Virginia §11-21-10 (b)
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/wv/tax/income/subtractions/low_income_earned_income/income_limit.yaml:7]:
```suggestion
    - title: Code of West Virginia §11-21-10 (b) (1) & (2)
```

File: policyengine_us/parameters/gov/states/wv/tax/income/subtractions/low_income_earned_income/amount.yaml [hunk 20]
```diff
@@ -0,0 +1,25 @@
+description: West Virginia caps the low-income earned income exclusion to this amount of earned income. 
+metadata:
+  period: year
+  unit: currency-USD  
+  label: West Virginia low-income earned income exclusion low-income earned income exclusion limit   
+  reference:
+    - title: Code of West Virginia §11-21-10 (b)
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/wv/tax/income/subtractions/low_income_earned_income/amount.yaml:7]:
```suggestion
    - title: Code of West Virginia §11-21-10 (a)
```

**Review by MaxGhenis — APPROVED** [review]

**Pull request merged**